explorer/interpreter/resolve_unformed.cpp

```diff
@@ -52,39 +68,52 @@ auto FlowFacts::TakeAction(Nonnull<const AstNode*> node, ActionType action,
     case ActionType::None:
       break;
   }
+
+  if (trace_stream_->is_enabled()) {
+    *trace_stream_ << "--- " << action_type_string(action) << " `" << name
+                   << "` (" << source_loc << ")\n";
+  }
+
   return Success();
 }
 
-static auto ResolveUnformedImpl(Nonnull<const Expression*> expression,
+static auto ResolveUnformedImpl(Nonnull<TraceStream*> trace_stream,
+                                Nonnull<const Expression*> expression,
                                 FlowFacts& flow_facts,
                                 FlowFacts::ActionType action)
     -> ErrorOr<Success>;
-static auto ResolveUnformedImpl(Nonnull<const Pattern*> pattern,
+static auto ResolveUnformedImpl(Nonnull<TraceStream*> trace_stream,
+                                Nonnull<const Pattern*> pattern,
                                 FlowFacts& flow_facts,
                                 FlowFacts::ActionType action)
     -> ErrorOr<Success>;
-static auto ResolveUnformedImpl(Nonnull<const Statement*> statement,
-                                FlowFacts& flow_facts,
-                                FlowFacts::ActionType action)
-    -> ErrorOr<Success>;
-static auto ResolveUnformedImpl(Nonnull<const Expression*> expression,
+static auto ResolveUnformedImpl(Nonnull<TraceStream*> trace_stream,
+                                Nonnull<const Statement*> statement,
                                 FlowFacts& flow_facts,
                                 FlowFacts::ActionType action)
     -> ErrorOr<Success>;
 
 // Traverses the sub-AST rooted at the given node, resolving the formed/unformed
 // states of local variables within it and updating the flow facts.
 template <typename T>
-static auto ResolveUnformed(Nonnull<const T*> expression, FlowFacts& flow_facts,
+static auto ResolveUnformed(Nonnull<TraceStream*> trace_stream,
+                            Nonnull<const T*> expression, FlowFacts& flow_facts,
                             FlowFacts::ActionType action) -> ErrorOr<Success> {
-  return RunWithExtraStack(
-      [&] { return ResolveUnformedImpl(expression, flow_facts, action); });
+  return RunWithExtraStack([&] {
+    return ResolveUnformedImpl(trace_stream, expression, flow_facts, action);
+  });
 }
 
-static auto ResolveUnformedImpl(Nonnull<const Expression*> expression,
+static auto ResolveUnformedImpl(Nonnull<TraceStream*> trace_stream,
+                                Nonnull<const Expression*> expression,
                                 FlowFacts& flow_facts,
                                 FlowFacts::ActionType action)
     -> ErrorOr<Success> {
+  if (trace_stream->is_enabled()) {
```

Comment:
I worry that this is going to produce a lot of output; is it reasonable to do the same thing we did for name resolution and only log statements and declarations?

Reply:
Yes, I think that sounds reasonable for this as well. I'll do the changes in the next commit.